sycl/include/CL/sycl/pipes.hpp

```diff
@@ -12,11 +12,33 @@
 #include <CL/__spirv/spirv_types.hpp>
 #include <CL/sycl/stl.hpp>
 
+#ifndef __SYCL_DEVICE_ONLY__
+#include <iostream>
+#endif // __SYCL_DEVICE_ONLY__
+
 namespace cl {
 namespace sycl {
 
-template <class name, class dataT, int32_t min_capacity = 0> class pipe {
+// Utility to dump typeid
+template <typename T> class wrap {};
+template <typename Type>
+std::string __str_type_name() {
+  return std::string(typeid(wrap<Type>).name());
+}
+
+template <typename PipeName>
+inline void __pipe_printer(std::string Input, int32_t S, int32_t A, int32_t C) {
+  std::cout << Input << __str_type_name<PipeName>() << std::endl
```

Comment:
Can you elaborate what is the use-case of this? Since pipes are by design known at compile time, it should be obvious what pipe a particular code line uses.

Reply:
Something like that:
`
using PipeA = cl::sycl::pipe<class A, int, 1>;
using PipeB = cl::sycl::pipe<class B, int, 1>;

int Key = { /*some calculations, may be even a data coming from another pipe takes place*/ };
if (Key % 2)
  PipeA::write(42);
else
  PipeB::write(42);
`

So running the program on host device will help to understand which pipe was used to write the data. It might be helpful, if we plan to read from a pipe, because if we executes a blocking read from a wrong one - the program will hang.

Reply:
> Something like that:
> 
> ```
> using PipeA = cl::sycl::pipe<class A, int, 1>;
> using PipeB = cl::sycl::pipe<class B, int, 1>;
> 
> int Key = { /*some calculations, may be even a data coming from another pipe takes place*/ };
> if (Key % 2)
>   PipeA::write(42);
> else
>   PipeB::write(42);
> ```
> 
> So running the program on host device will help to understand which pipe was used to write the data. It might be helpful, if we plan to read from a pipe, because if we executes a blocking read from a wrong one - the program will hang.

I see. Is it possible to subclass a pipe and add pre/post hooks for reads and writes?
In other words, can a user write its own instrumentation or logging for pipes?

Reply:
To be more specific: if the following code works, user is free to add any debugging facilities he needs; no need to patch the headers - we cannot provide a solution that is sufficient in any debugging scenario anyway. It is always better to let a user define these things.

```
template<param1, param2>
class debug_pipe : public pipe<param1, param2> {
  void write(int i) {
    printf("debug_pipe writes %d\n", i);
    pipe<param1, param2>::write(i);
  }
}

using PipeA = debug_pipe<int, ...>;
using PipeB = debug_pipe<int,...>
```

Reply:
Lets close the PR then. My original plan was to use __str_type_name in a future development, so the patch was just built around it.